osquery/core/database/in_memory_database.cpp

```diff
@@ -0,0 +1,177 @@
+         << domain;
+}
+
+template <typename T>
+Expected<T, DatabaseError> InMemoryDatabase::getValue(const std::string& domain,
+                                                      const std::string& key) {
+  assert(is_open_ && "database is not open");
+  if (!is_open_) {
+    return createError(DatabaseError::DatabaseIsNotOpen, "Database is closed");
+  }
+  auto storage_iter = storage_.find(domain);
+  if (storage_iter == storage_.end()) {
+    return domainNotFoundError(domain);
+  }
+  std::lock_guard<std::mutex> lock(storage_iter->second->getMutex());
+  auto result = storage_iter->second->get(key);
+  if (result) {
+    DataType value = result.take();
+    if (value.type() == typeid(T)) {
+      return boost::get<T>(value);
+    } else {
+      auto error =
+          createError(DatabaseError::KeyNotFound, "Requested wrong type for: ")
+          << domain << ":" << key << " stored type: " << value.type().name()
+          << " requested type " << boost::core::demangle(typeid(T).name());
+      LOG(ERROR) << error.getFullMessageRecursive();
+      debug_only::fail(error.getFullMessageRecursive().c_str());
+      return std::move(error);
+    }
+  }
+  return result.takeError();
+}
+
+template <typename T>
+ExpectedSuccess<DatabaseError> InMemoryDatabase::putValue(
+    const std::string& domain, const std::string& key, const T& value) {
+  assert(is_open_ && "database is not open");
+  if (!is_open_) {
+    return createError(DatabaseError::DatabaseIsNotOpen, "Database is closed");
+  }
+  auto storage_iter = storage_.find(domain);
+  if (storage_iter == storage_.end()) {
+    return domainNotFoundError(domain);
+  }
+  std::lock_guard<std::mutex> lock(storage_iter->second->getMutex());
+  debug_only::verify(
+      [&storage_iter, &key]() {
+        auto result = storage_iter->second->get(key);
+        return result && result.get().type() == typeid(T) &&
+               "changing type is not allowed";
+      },
+      "changing type is not allowed");
+  storage_iter->second->put(key, value);
+  return Success();
+}
+
+Expected<std::string, DatabaseError> InMemoryDatabase::getString(
+    const std::string& domain, const std::string& key) {
+  return getValue<std::string>(domain, key);
+}
+
+ExpectedSuccess<DatabaseError> InMemoryDatabase::putString(
+    const std::string& domain,
+    const std::string& key,
+    const std::string& value) {
+  return putValue(domain, key, value);
+}
+
+Expected<int, DatabaseError> InMemoryDatabase::getInt32(
+    const std::string& domain, const std::string& key) {
+  return getValue<int32_t>(domain, key);
+}
+
+ExpectedSuccess<DatabaseError> InMemoryDatabase::putInt32(
+    const std::string& domain, const std::string& key, const int32_t value) {
+  return putValue(domain, key, value);
+}
+
+Expected<std::vector<std::string>, DatabaseError> InMemoryDatabase::getKeys(
+    const std::string& domain, const std::string& prefix) {
+  assert(is_open_ && "database is not open");
```

Comment:
Just a suggestion, you could use `debug_only::verifyTrue(is_open_, "...");` here. Maybe you will find this way more readable :)